pkg/stanza/fileconsumer/internal/reader/compressedreader.go

```diff
@@ -0,0 +1,147 @@
+package reader
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/header"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/scanner"
+	"go.uber.org/zap"
+	"golang.org/x/net/context"
+	"math"
+	"os"
+)
+
+type CompressedReader struct {
+	*Reader
+
+	buffer           *bytes.Buffer
+	lastReadFileSize int64
+	fromBeginning    bool
+}
+
+func (r *CompressedReader) ReadToEnd(ctx context.Context) {
+	newFileHandle, err := os.Open(r.file.Name())
+	if err != nil {
+		r.set.Logger.Error("problem creating reader for compressed file", zap.Error(err))
+		return
+	}
+	r.file = newFileHandle
+	gzipReader, err := gzip.NewReader(r.file)
+	if err != nil {
+		r.set.Logger.Error("problem creating reader for compressed file", zap.Error(err))
+		return
+	}
+
+	// TODO validate if we can keep this check for the file size to avoid unnecessary decompression
+	// in theory, this check can lead to new log entries going unnoticed, if the new entries are duplicates
+	// of a previous entry, as this leads to the size of the compressed file not being increased.
+	// practically this may not be an issue for logs, since at least the timestamp should change for a new log entry,
+	// even if it has the same message as the line before
+
+	stat, err := r.file.Stat()
+	if err != nil {
+		r.set.Logger.Error("problem creating determining size of compressed file", zap.String("fileName", r.fileName), zap.Error(err))
+		return
+	}
+
+	initalRead := true
+	if lastReadFileSize, ok := r.FileAttributes["lastReadFileSize"]; ok {
+		initalRead = false
+		if lastReadFileSize.(int64) >= stat.Size() {
+			// do not read further unless something has been added to the file
+			return
+		}
+	}
+
+	r.FileAttributes["lastReadFileSize"] = stat.Size()
+
+	uncompressedBytes := make([]byte, r.maxLogSize)
+	nrBytes, err := gzipReader.Read(uncompressedBytes)
+
+	scannedBytes := int64(0)
+	defer func() {
+		if r.needsUpdateFingerprint {
+			r.updateFingerprint()
+		}
+		if !r.fromBeginning && initalRead {
```

Comment:
In the other reader, the `fromBeginning` flag is set to false after the initial read. Can we just do that here too, to avoid having to manage the extra parameter? Moreover, this may be another part of the code that can be shared unless there's a specific reason the implementation should be different.

Reply:
i think we can do that, yes - I think i slightly misinterpreted the semantics of that parameter as i thought this is referring to the `start_at` config option of the filelog receiver and is always set to the same value. Thank you for clarifying

Reply:
No problem. It's not super clear, but `fromBeginning` is initialized according to `start_at`, but then updated to `true` after the first poll. Basically, `start_at` only applies to files which exist prior to starting the collector. Any new files are understood to contain new logs so we should read them entirely.